Datadog Integrations

blarghmatey opened the pull request:
This pull request will add the ability to define configuration parameters for Datadog and for its plugins via pillar data. It also includes states for configuring the integrations with MongoDB and MySQL.

closes mitodl/salt-ops#29

blarghmatey committed:
Added ability to write config via pillar

Updated the formula to install the DataDog agent and write out a valid
config file via data contained in the `datadog:config` path in pillar data.

blarghmatey committed:
Added integration configuration

- Added the ability to define integration settings via pillar data.
- Added setup states for Mongo and MySQL integrations
- Added tests to validate presence of plugin configs

bdero commented on datadog/plugins/init.sls:
```diff
@@ -0,0 +1,20 @@
+  file.directory:
+    - name: /etc/dd-agent/conf.d
+
+{% for name, settings in salt.pillar.get('datadog:integrations', {}).items() %}
```
I know this works, but is there any reason to not use the merged `datadog` context rather than the pillar data directly?

blarghmatey replied:
No, good catch. I had originally had the map file merging in the `datadog:lookups` key path which is why I had gone with the pillar approach at first. Now that it's just merging the whole datadog key path I could go with the map data. That being said, it's an open question as to whether it's a good idea to merge in the whole keyspace to the map file or if it should just be a subset of that data. Maybe make the convention be `<formula_name>:overrides` instead of `<formula_name>:lookups`. Thoughts?

bdero replied:
As we discussed in person, `overrides` for a name might be better, and merging the root key in can cause confusion and obscure the separation between forced pillar data entry and overridable defaults, so the convenience benefit probably doesn't outweigh the disadvantages.

bdero commented on datadog/plugins/mongodb.sls:
```diff
@@ -0,0 +1,18 @@
+    'datadog:integrations:mongodb',
+    {
+        'passwd': 'Ch@ng3me!',
+        'password': 'N0tS3cure!'
```
I know 'passwd' and 'password' are the names of the params for `mongodb_user.present`, but can we use something more descriptive for the var names? Perhaps `admin_password` and `datadog_password`.

blarghmatey replied:
Valid point, I'll do that.

bdero commented on datadog/plugins/mongodb.sls:
```diff
@@ -0,0 +1,18 @@
+    {
+        'passwd': 'Ch@ng3me!',
+        'password': 'N0tS3cure!'
+    }) %}
```
It seems worth adding entries for these in the `pillar.example`, especially the one corresponding with the datadog "passwd" - since it makes it harder for someone to accidentally only add their admin "password" and wind up keeping the other default.

blarghmatey replied:
Good call

bdero commented on datadog/plugins/mysql.sls:
```diff
@@ -0,0 +1,13 @@
+    'datadog:integrations:mysql',
+    {
+        'password': 'Ch@ng3me!',
+        'host': '*.*',
```
Similar comment as mongo: Can entries for these be added to `pillar.example`, even if it has to be commented out by default?

bdero commented on datadog/plugins/mysql.sls:
```diff
@@ -0,0 +1,13 @@
+    - name: datadog
+    - password: {{ dd_mysql.password }}
+    - host: {{ dd_mysql.host }}
+    {% for }
```
It looks like line is an unfinished jinja2 fragment

blarghmatey replied:
Good catch

bdero commented:
:tada: After comments addressed, :+1:

blarghmatey committed:
Updated attribute names for mongo/mysql credentials

- Updated the attribute names for mongo and mysql credentials in plugin setup
- Updated pillar path to use overrides for merging with map data
- Updated pillar example to include keys for mongo and mysql integrations

bdero commented:
:+1:
